lib/actions/preload.coffee

```diff
@@ -0,0 +1,208 @@
+###
+Copyright 2016-2017 Resin.io
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+###
+
+_ = require('lodash')
+form = require('resin-cli-form')
+preload = require('resin-preload')
+Promise = require('bluebird')
+resin = require('resin-sdk-preconfigured')
+
+dockerUtils = require('../utils/docker')
+
+LATEST = 'latest'
+
+selectApplication = (resin, form, deviceType) ->
+	resin.models.application.getAll({filter: {device_type: deviceType}})
+	.then (applications) ->
+		if applications.length == 0
+			expectedError("You have no apps for a '#{deviceType}' device type.")
+		form.ask
+			message: 'Select an application'
+			type: 'list'
+			choices: applications.map (app) ->
+				name: app.app_name
+				value: app
+
+selectApplicationCommit = (resin, form, appId) ->
+	resin.models.build.getAllByApplication(appId)
+	.then (builds) ->
+		if builds.length == 0
+			expectedError('This application has no builds')
+		builds = builds.filter (build) ->
+			build.status == 'success'
+		builds = _.sortBy(builds, '-push_timestamp')
```

Comment:
Is it OK if you find no successful builds? Probably not too

Also why not filtering and ordering on the server side (using pine options)?